[lit] Fix some issues from --per-test-coverage

jdenny-ornl opened the pull request:
Key issue
---------

When lit is configured to use windows `cmd` as an external shell, it appears that all RUN lines are effectively commented out.  The problematic change landed in July 26, 2023, so it's surprising it hasn't come up yet.  I'm afraid I don't have a windows setup, so I could use some help to verify what's happening.

Details
-------

In the case that lit is configured to use an external shell, D154280 (landed in 64d19542e78a on July 26, 2023) causes
`%dbg(RUN: at line N)` to be expanded in RUN lines early and in a manner that is specific to sh-like shells.  As a result, later code in lit that expands it in a shell-specific manner is useless.

That sh-like expansion uses the `:` command as follows:

```
: 'RUN: at line N'; original-commands
```

In sh-like shells, the `:` command simply discards its arguments. However, in windows `cmd`, `:` indicates a goto label.  That appears to effectively comment out the rest of the line, including the original commands from the RUN line.

I am not aware of any complaints about this change.  Did I miss them? Are all tests still passing and so no one noticed?  Lit's own test suite has some tests that normally fail if RUN lines don't execute. Is no one running lit's test suite with windows `cmd` as a lit external shell?  Or is no one using windows `cmd` as a lit external shell at all anymore?

Another issue
-------------

D154280 doesn't implement `--per-test-coverage` for lit's internal shell.

Fix
---

This patch fixes the above problems by implementing `--per-test-coverage` before selecting the shell (internal or external) and by leaving `%dbg(RUN: at line N)` unexpanded.  Thus, it is expanded later in a shell-specific manner, as before D154280.

I would like to understand whether windows `cmd` as a lit external shell is worthwhile to support anymore.

jh7370 commented:
What is `%dbg` actually used for? I couldn't find any references to it in the main LLVM tests, nor is there any reference to it in the Testing Guid docs, so I suspect I'm missing something crucial to understanding this situation...

> I would like to understand whether windows `cmd` as a lit external shell is worthwhile to support anymore.

We probably should, purely on the basis that people who develop on Windows might be using this setup. That being said, I'm fairly confident that Windows users by default use the internal shell. I have not verified that though.

mstorsjo commented:
> > I would like to understand whether windows `cmd` as a lit external shell is worthwhile to support anymore.
> 
> We probably should, purely on the basis that people who develop on Windows might be using this setup. That being said, I'm fairly confident that Windows users by default use the internal shell. I have not verified that though.

FWIW, I'm not a regular Windows user, I only run it in VMs and similar for testing things, but whenever I do and want to test things in a native Windows environment (as opposed to running in msys2 bash or similar), I do it with cmd. I've never really gotten familiar with powershell.

And whenever tools like python etc, programmatically execute some command string in a way that it is parsed by a shell, that shell is cmd, not powershell afaik.

mstorsjo commented:
I haven't been so much involved in the recent Lit changes, and it's been a while since I poked at these things, so it took me a little bit of fiddling around to get back on track about what the actual status quo is here. I'll try to summarize my findings, please doublecheck if I've missed something:

Status quo
---
- `ShTest` defaults to `execute_external=False` (i.e., internal shell) if no parameter has been passed: https://github.com/llvm/llvm-project/blob/llvmorg-18-init/llvm/utils/lit/lit/formats/shtest.py#L22
- The LLVM testsuite initializes this based on `llvm_config.use_lit_shell` - https://github.com/llvm/llvm-project/blob/llvmorg-18-init/llvm/test/lit.cfg.py#L21. The vast majority of testsuites in LLVM also copies this behaviour.
- The LLVM testsuite defaults to `self.use_lit_shell = False`, except that on Windows it defaults to `self.use_lit_shell = True`. This can be overridden with the environment variable `LIT_USE_INTERNAL_SHELL`. https://github.com/llvm/llvm-project/blob/llvmorg-18-init/llvm/utils/lit/lit/llvm/config.py#L23-L52 I.e. if people don't explicitly opt in to it by setting `LIT_USE_INTERNAL_SHELL=0`, people will be using the internal shell on Windows
- The OpenMP testsuite initializes `ShTest` with the defaults, i.e. always using the internal shell: https://github.com/llvm/llvm-project/blob/llvmorg-18-init/openmp/runtime/test/lit.cfg#L42
- compiler-rt's tests have custom setup logic that doesn't reuse the same setup as most other tools with `llvm_config.use_lit_shell`, but has essentially equivalent logic here: https://github.com/llvm/llvm-project/blob/llvmorg-18-init/compiler-rt/test/lit.common.cfg.py#L113-L128
- The libcxx testsuite (which is reused for libcxxabi and libunwind) defaults to the internal shell, without any option for overriding it: https://github.com/llvm/llvm-project/blob/llvmorg-18-init/libcxx/utils/libcxx/test/format.py#L384-L387
- If using an external shell for running tests, this will be bash, if bash was found in the path - if not, it will be the win32 cmd.exe: https://github.com/llvm/llvm-project/blob/llvmorg-18-init/llvm/utils/lit/lit/TestRunner.py#L1068-L1069

As additional context: The libcxx testsuite used to use the external shell a few years ago, but this was problematic for running the tests on Windows, it didn't really work right there (I don't have exact details). This was changed in 39bbfb77264a4a7a216921c2b70a30ba0f27eb56 - before that, running the libcxx tests required making sure that `bash` was available in path.

I.e., unless jumping through hoops, none of these testsuites really run things with the win32 cmd.exe.

Breakage
---
Running things with win32 cmd.exe as external shell did use to work, but currently doesn't. This actually seems to have broken in 1041a9642ba035fd2685f925911d705e8edf5bb0 already. If running a simple llvm shell test with `LIT_USE_INTERNAL_SHELL=0`, before that, I got a successful run:
```
Script:
--
echo 'RUN: at line 3' > nul &&   c:\code\llvm-project\llvm\build\bin\llvm-mc.exe -triple aarch64-pc-win32 -filetype=obj C:/code/llvm-project/llvm/test/MC/AArch64/seh.s | c:\code\llvm-project\llvm\build\bin\llvm-readobj.exe -S -r -u - | c:\code\llvm-project\llvm\build\bin\filecheck.exe C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
echo 'RUN: at line 8' > nul &&   c:\code\llvm-project\llvm\build\bin\llvm-mc.exe -triple aarch64-pc-win32 -filetype=asm C:/code/llvm-project/llvm/test/MC/AArch64/seh.s | c:\code\llvm-project\llvm\build\bin\llvm-mc.exe -triple aarch64-pc-win32 -filetype=obj - | c:\code\llvm-project\llvm\build\bin\llvm-readobj.exe -S -r -u - | c:\code\llvm-project\llvm\build\bin\filecheck.exe C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
--
Exit Code: 0
```

After that commit, it instead breaks like this:
```
Script:
--
echo 'RUN: at line 3' > nul &&
echo 'RUN: at line 8' > nul &&
--
Exit Code: 255

Command Output (stderr):
--
The syntax of the command is incorrect.
```

It further breaks at 64d19542e78a43edb7ae26ea6762a2b1c360a916 as you've noticed, since the RUN lines essentially just become no-op goto labels.
```
Script:
--
: 'RUN: at line 3';   c:\code\llvm-project\llvm\build\bin\llvm-mc.exe -triple aarch64-pc-win32 -filetype=obj C:/code/llvm-project/llvm/test/MC/AArch64/seh.s | c:\code\llvm-project\llvm\build\bin\llvm-readobj.exe -S -r -u - | c:\code\llvm-project\llvm\build\bin\filecheck.exe C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
: 'RUN: at line 8';   c:\code\llvm-project\llvm\build\bin\llvm-mc.exe -triple aarch64-pc-win32 -filetype=asm C:/code/llvm-project/llvm/test/MC/AArch64/seh.s | c:\code\llvm-project\llvm\build\bin\llvm-mc.exe -triple aarch64-pc-win32 -filetype=obj - | c:\code\llvm-project\llvm\build\bin\llvm-readobj.exe -S -r -u - | c:\code\llvm-project\llvm\build\bin\filecheck.exe C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
--
Exit Code: 0
```
(It prints this even if injecting an error in the test.)


Furthermore, 09b6e457d91ce84088e6e21783913e5f1e5bd227 changes this case further, making it just print this:
```
PASS: LLVM :: MC/AArch64/seh.s (1 of 1)
Exit Code: 0
```
(Even when executing with `-a -v -v`.)


Separate side note; running the LLVM testsuite with bash as external executor (if it happens to be available in path) also seems to fail:
```
Exit Code: 127

Command Output (stderr):
--
+ : 'RUN: at line 3'
+ c:codellvm-projectllvmbuildbinllvm-mc.exe -triple aarch64-pc-win32 -filetype=obj C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
C:\code\\llvm-project\llvm\build\test\MC\AArch64\Output\seh.s.script: line 1: c:codellvm-projectllvmbuildbinllvm-mc.exe: command not found
+ c:codellvm-projectllvmbuildbinllvm-readobj.exe -S -r -u -
C:\code\\llvm-project\llvm\build\test\MC\AArch64\Output\seh.s.script: line 1: c:codellvm-projectllvmbuildbinllvm-readobj.exe: command not found
+ c:codellvm-projectllvmbuildbinfilecheck.exe C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
C:\code\\llvm-project\llvm\build\test\MC\AArch64\Output\seh.s.script: line 1: c:codellvm-projectllvmbuildbinfilecheck.exe: command not found

--
```
(This seems to have been broken throughout this time period at least. However before 39bbfb77264a4a7a216921c2b70a30ba0f27eb56, when libcxx tests were using the external executor, they did work when executed with bash as external executor.)

Conclusion
---
It certainly looks like this configuration is severely broken. I'm not aware of any parts of the configurations that use it by default, and even then, it's generally problematic, so I'm not sure if it's worth spending effort on fixing. (The suggestion that was made somewhere around these discussions, to move most things towards using the internal shell even on unix, sounds to me like a good direction in general .)

That said; for echoing executed commands, when using the internal executor, it's probably very valuable if they're printed in a form that is easy to copypaste and execute in cmd.exe (where possible - at least for the majority of simple cases).

mstorsjo commented:
FWIW, this PR doesn't entirely fix running the tests with win32 cmd.exe either; with this PR, it fails like this:
```
Exit Code: 255

Command Output (stdout):
--

C:\code\llvm-project\llvm\build\test\MC\AArch64>echo 'RUN: at line 3' > nul &&

--
Command Output (stderr):
--
The syntax of the command is incorrect.

```

hnrklssn commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -2115,10 +2100,35 @@ def parseIntegratedTestScript(test, additional_parsers=[], require_script=True):
+                coverage_index += 1
+                command = f"export LLVM_PROFILE_FILE={profile} && {command}"
+                if match:
+                    command = buildPdbgCommand(dbg, command)
```
This looks correct to me, but it would be good to have a regression test for it.

jdenny-ornl replied:
This PR does extend lit's test suite to cover more cases, like lit's internal shell.  What coverage are you looking for?

hnrklssn replied:
The `%dbg` case

jdenny-ornl replied:
In case it's not clear, `%dbg` is normally used in lit.  It produces the `RUN: at line N` info.

It's the windows `cmd` use case that appears to be uncommon (or non-existent now), and `%dbg` was broken for that case.  I think that case will be checked by `execute_external=True` tests in lit's test suite... if anyone actually runs lit's test suite such that the external shell is windows `cmd`.

Even if we eliminate support for windows `cmd` as a lit external shell, I think this patch will live on in some form to remove unreachable `%dbg` expansion code and to support `--per-test-coverage` with lit's internal shell.  With that in mind, I added a commit to extend test coverage for `%dbg` behavior when `--per-test-coverage` is used.

hnrklssn replied:
Yes, that's exactly the type of test I had in mind, thanks! LGTM

jdenny-ornl commented:
> What is `%dbg` actually used for? I couldn't find any references to it in the main LLVM tests, nor is there any reference to it in the Testing Guid docs, so I suspect I'm missing something crucial to understanding this situation...

It's something like an internal substitution.  When parsing a RUN line, [lit inserts it before the RUN line's commands](https://github.com/llvm/llvm-project/blob/2f8690b1e2bf09174660cd66ef8a27066848555e/llvm/utils/lit/lit/TestRunner.py#L1913) like this:

```
%dbg(RUN: at line N) commands
```

Later, lit expands it based on the shell (internal, external sh-like shell, external windows `cmd`, etc.).  For example, [for sh-like shells](https://github.com/llvm/llvm-project/blob/2f8690b1e2bf09174660cd66ef8a27066848555e/llvm/utils/lit/lit/TestRunner.py#L1186-L1189):

```
: 'RUN: at line N'; commands
```

jdenny-ornl commented:
> I haven't been so much involved in the recent Lit changes, and it's been a while since I poked at these things, so it took me a little bit of fiddling around to get back on track about what the actual status quo is here. I'll try to summarize my findings, please doublecheck if I've missed something:

Thanks for the nice summary.  It looks right to me.

> Running things with win32 cmd.exe as external shell did use to work, but currently doesn't. This actually seems to have broken in [1041a96](https://github.com/llvm/llvm-project/commit/1041a9642ba035fd2685f925911d705e8edf5bb0) already.

You're right, and that landed in April, 2022, so windows `cmd` support has been broken a while.  That part is the reason why the current PR wasn't sufficient to restore correct behavior, as you point out later in your comment.  It looks straight-forward to fix, but maybe doing so isn't worthwhile.
 
> Separate side note; running the LLVM testsuite with bash as external executor (if it happens to be available in path) also seems to fail:
> 
> ```
> Exit Code: 127
> 
> Command Output (stderr):
> --
> + : 'RUN: at line 3'
> + c:codellvm-projectllvmbuildbinllvm-mc.exe -triple aarch64-pc-win32 -filetype=obj C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
> C:\code\\llvm-project\llvm\build\test\MC\AArch64\Output\seh.s.script: line 1: c:codellvm-projectllvmbuildbinllvm-mc.exe: command not found
> + c:codellvm-projectllvmbuildbinllvm-readobj.exe -S -r -u -
> C:\code\\llvm-project\llvm\build\test\MC\AArch64\Output\seh.s.script: line 1: c:codellvm-projectllvmbuildbinllvm-readobj.exe: command not found
> + c:codellvm-projectllvmbuildbinfilecheck.exe C:/code/llvm-project/llvm/test/MC/AArch64/seh.s
> C:\code\\llvm-project\llvm\build\test\MC\AArch64\Output\seh.s.script: line 1: c:codellvm-projectllvmbuildbinfilecheck.exe: command not found
> 
> --
> ```
> 
> (This seems to have been broken throughout this time period at least. However before [39bbfb7](https://github.com/llvm/llvm-project/commit/39bbfb77264a4a7a216921c2b70a30ba0f27eb56), when libcxx tests were using the external executor, they did work when executed with bash as external executor.)

It looks like the `llvm-mc` substitution isn't quoting its value correctly.

> ## Conclusion
> 
> It certainly looks like this configuration is severely broken. I'm not aware of any parts of the configurations that use it by default, and even then, it's generally problematic, so I'm not sure if it's worth spending effort on fixing. (The suggestion that was made somewhere around these discussions, to move most things towards using the internal shell even on unix, sounds to me like a good direction in general .)

Agreed.  What if we raise a python exception on the windows `cmd` code path in lit for now?  If we still see no complaints for a few weeks, we can think about removing the code path entirely.

> That said; for echoing executed commands, when using the internal executor, it's probably very valuable if they're printed in a form that is easy to copypaste and execute in cmd.exe (where possible - at least for the majority of simple cases).

If RUN lines are written for lit's internal shell, I'm not sure if they are generally guaranteed to work in windows `cmd`.  Do you have a specific change in mind here?

mstorsjo commented:
> > (This seems to have been broken throughout this time period at least. However before [39bbfb7](https://github.com/llvm/llvm-project/commit/39bbfb77264a4a7a216921c2b70a30ba0f27eb56), when libcxx tests were using the external executor, they did work when executed with bash as external executor.)
> 
> It looks like the `llvm-mc` substitution isn't quoting its value correctly.

Yep, probably - I guess this goes the same for all the llvm tools that have substitutions set up.

> > ## Conclusion
> > It certainly looks like this configuration is severely broken. I'm not aware of any parts of the configurations that use it by default, and even then, it's generally problematic, so I'm not sure if it's worth spending effort on fixing. (The suggestion that was made somewhere around these discussions, to move most things towards using the internal shell even on unix, sounds to me like a good direction in general .)
> 
> Agreed. What if we raise a python exception on the windows `cmd` code path in lit for now? If we still see no complaints for a few weeks, we can think about removing the code path entirely.

I guess that sounds reasaonble. Although I'm not sure how often odd downstreams integrate from upstream, but things have been quite broken for quite some time anyway, so I guess it's pretty safe to say that this is dead code.

> > That said; for echoing executed commands, when using the internal executor, it's probably very valuable if they're printed in a form that is easy to copypaste and execute in cmd.exe (where possible - at least for the majority of simple cases).
> 
> If RUN lines are written for lit's internal shell, I'm not sure if they are generally guaranteed to work in windows `cmd`. Do you have a specific change in mind here?

Not really, no - I'm mostly commenting on it from the point of view that I remember seeing somewhere else in these discussions (regarding formatting of the `echo RUN` stuff I think). As long as most commands, in the form `command1 | command2` work and can be copypasted as such, we're probably just fine. (And I have no reason to believe that wouldn't be the case.)

RoboTux reviewed (dismissed):
LGTM

jdenny-ornl commented:
Thanks for everyone's comments so far.

This PR is not ready to land.  As discussed, it doesn't fully fix support for windows `cmd` as a lit external shell.  Given how long that use case has been broken (apparently since April, 2022), it seems that's a use case no one cares about anymore.

Possible next steps include:

1. Fix windows `cmd` support.  That looks straight-forward.  If we're going to do that, then we probably should also apply changes that PR #65267 applied for sh-like external shells, and there [might still be further changes along those lines as well](https://discourse.llvm.org/t/rfc-improving-lits-debug-output/72839/44?u=jdenny-ornl).  But I don't see a reason to go down this path if no one actually cares.
2. Land a patch to raise a python exception when lit encounters its code path to use windows `cmd` as an external shell.  If no one complains for a few weeks, then we drop support.
3. Rewrite this PR's title and commit log to explain what it does do: clean up the `%dbg` expansion implementation, and make `--per-test-coverage` work for lit's internal shell.

Please let me know if you have opinions on this.  When I have some more time, I'm thinking of pursuing 1 and 3.

Endilll commented:
Whatever we do, I consider it important for us to understand how widely `cmd` is used.

compnerd commented:
I think that might be a bit too accelerated.  I work mostly with a fork of LLVM (https://github.com/apple/llvm-project) and do care about the usability of `cmd` as a shell.  I think that Swift is a large enough consumer of LLVM to consider testing that before declaring this feature removed.

jdenny-ornl commented:
> I think that might be a bit too accelerated. I work mostly with a fork of LLVM (https://github.com/apple/llvm-project) and do care about the usability of `cmd` as a shell (however, it is significantly behind but is working on updating to a more recent version). I think that Swift is a large enough consumer of LLVM to consider testing that before declaring this feature removed.

Thanks for mentioning this.

Given that `cmd` as a lit external shell appears to be have been unusable since April, 2022, how is that fork relying on it?  Is it farther behind than that?

Fixing the remaining problem with `cmd` appears to be straight-forward, so I'm fine to do that.  However, how hard would it be to use lit's internal shell for your use case?

Endilll commented:
> When I install Visual Studio, it creates a widget that brings up a cmd shell with the right environment. I'm not aware of a similar widget for PowerShell.

At least in VS 2022, there is a `Developer PowerShell for VS 2022` in Start menu beside the usual `x64 Native Tools Command Prompt for VS 2022`. It was there in 2019, but IIRC dropped you into x86 toolchain.

jdenny-ornl commented:
> I have no stake in Apple/Swift, but for myself, `cmd` is the natural shell to use on Windows. (When I install Visual Studio, it creates a widget that brings up a `cmd` shell with the right environment. I'm not aware of a similar widget for PowerShell.) While I rarely try to copy-paste from a `lit` dump into `cmd` (the commands are generally too full of long paths to temp locations to work well) I think it would be good (but not mandatory) for that copy-paste to actually work.
> 
> Like others have said, I had not really grokked that Windows defaults to the internal shell, although just this week I tripped over a behavioral difference in how `diff` works in the two environments (not related to your patches). So, if there seems to be benefit to making the external shell Just Not Work on Windows, I wouldn't lose a lot of sleep over it.

In case it isn't clear, there are two somewhat orthogonal issues here:

- This PR is about whether lit can execute RUN lines in `cmd`.  That does not appear to have worked since April, 2022. 
- The copy-paste issue is about whether lit can print RUN lines in a form that the user can easily copy to `cmd`.  For RUN lines that were written for lit's internal shell, that might require lit to *translate* the RUN lines to a form compatible with `cmd`.  I say might because I don't have much experience there.  The `diff` example above sounds like an example.

jdenny-ornl commented:
>     3. Rewrite this PR's title and commit log to explain what it does do: clean up the `%dbg` expansion implementation, and make `--per-test-coverage` work for lit's internal shell.

I just modified this PR accordingly and squashed previous commits into a single commit.

In the new commit log, I've summarized the windows `cmd` situation, which this PR improves but does not finish repairing.  That support can be fully repaired or discarded in a separate PR after we have more information on lit users' current needs.

I've also incorporated the latest changes from main into this PR.  To do so, I rebased and force-pushed (instead of merging), following the advice I most recently saw from various discussions about LLVM PRs.

The resulting diff is different than before, primarily because of all my lit patches I had to revert recently.  I also extended the testing.  It's probably best if reviewers look over things again before I try to land this.

This patch has code cleanup that is helpful for my other lit patches, and it doesn't seem to be controversial, so I'm now planning to land it first.

tru commented:
I noticed that the python formatting check failed so I looked into that, the problem is the file `llvm/utils/lit/tests/per-test-coverage.py`

this file is not valid python, neither `black` or `python` can parse it. Since that's the case, can we change the extension on it so that we don't have to create a lot of exceptions in our CI and pre-commit hook?

jdenny-ornl commented:
> I noticed that the python formatting check failed so I looked into that, the problem is the file `llvm/utils/lit/tests/per-test-coverage.py`

Thanks for pointing that out.  I noticed it too but made the bad assumption that something was wrong with the pre-commit check itself, and I ran out of time to investigate.  This is the diagnostic I saw after a traceback:

```
black.parsing.InvalidInput: Cannot parse: 12:13: t-cfg.py ({{[^)]*}})
```

I didn't see `llvm/utils/lit/tests/per-test-coverage.py` mentioned anywhere.  Did I miss it?

In contrast, when I run black on the command line on my local system, I see:

```
$ black llvm/utils/lit/tests/per-test-coverage.py 
error: cannot format llvm/utils/lit/tests/per-test-coverage.py: Cannot parse: 12:13: t-cfg.py ({{[^)]*}})

Oh no! 💥 💔 💥
1 file failed to reformat.
```

Is there any way to get the pre-commit check to mention the name of the problematic file?

(By the way, I have noticed that many python files were formatted with >80 columns, probably the black default of 88.  That doesn't fit in the 80-column editor/terminal windows I use for most of LLVM.  Is there any good reason not to follow [the LLVM coding standard for source code width](https://llvm.org/docs/CodingStandards.html#source-code-width) in python files?  I see no exception there for python, or did I misunderstand something?  Black just needs a `-l80` command-line option.)

> 
> this file is not valid python, neither `black` or `python` can parse it. Since that's the case, can we change the extension on it so that we don't have to create a lot of exceptions in our CI and pre-commit hook?

Lit's tests use the `.py` suffix.  Most just contain comments that contain lit/FileCheck directives.  Some contain actual python code called by RUN lines.  The above test contains garbage outside of a comment, so the test works fine, but it's invalid as python.  I'll push a commit to this PR to fix it.  Thanks again for pointing me in the right direction.

tru commented:
> > I noticed that the python formatting check failed so I looked into that, the problem is the file `llvm/utils/lit/tests/per-test-coverage.py`
> 
> Thanks for pointing that out. I noticed it too but made the bad assumption that something was wrong with the pre-commit check itself, and I ran out of time to investigate. This is the diagnostic I saw after a traceback:
> Is there any way to get the pre-commit check to mention the name of the problematic file?
> 

I'll look into this, it also took me time to figure it out. 

> (By the way, I have noticed that many python files were formatted with >80 columns, probably the black default of 88. That doesn't fit in the 80-column editor/terminal windows I use for most of LLVM. Is there any good reason not to follow [the LLVM coding standard for source code width](https://llvm.org/docs/CodingStandards.html#source-code-width) in python files? I see no exception there for python, or did I misunderstand something? Black just needs a `-l80` command-line option.)
> 

This was discussed when we wrote the coding style for python and then we thought it was just better to use black's default to avoid having people need to run black with arguments (since it won't read a config file).

tru commented:
> > > (By the way, I have noticed that many python files were formatted with >80 columns, probably the black default of 88. That doesn't fit in the 80-column editor/terminal windows I use for most of LLVM. Is there any good reason not to follow [the LLVM coding standard for source code width](https://llvm.org/docs/CodingStandards.html#source-code-width) in python files? I see no exception there for python, or did I misunderstand something? Black just needs a `-l80` command-line option.)
> > 
> > 
> > This was discussed when we wrote the coding style for python and then we thought it was just better to use black's default to avoid having people need to run black with arguments
> 
> OK, but the LLVM coding standard appears to contradict that approach, as far as I can tell:
> 
> 1. "Write your code to fit within 80 columns."  (From https://llvm.org/docs/CodingStandards.html#source-code-width.)
> 2. "The Python code within the LLVM repository should adhere to the formatting guidelines outlined in [PEP-8](https://peps.python.org/pep-0008/)."  (From https://llvm.org/docs/CodingStandards.html#python-version-and-source-code-formatting).
> 3. "Limit all lines to a maximum of 79 characters." (From https://peps.python.org/pep-0008/#maximum-line-length.)
> 
> The [LLVM coding standard's advice on using black](https://llvm.org/docs/CodingStandards.html#python-version-and-source-code-formatting) is:
> 
> > For consistency and to limit churn, code should be automatically formatted with the [black](https://github.com/psf/black) utility. Black allows changing the formatting rules based on major version. In order to avoid unnecessary churn in the formatting rules we currently use black version 23.x in LLVM.
> 
> Combining all the above, I'm led to believe I should be passing options to black, but it's unclear which ones.
> 
> I'm sorry I didn't participate in the original discussion. If it's too late to change things, can we clarify the LLVM coding standard?
> 
> > (since it won't read a config file).
> 
> See https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#configuration-via-a-file.
> 
> For example, in `llvm/utils/lit/pyproject.toml `, I appended the following:
> 
> ```
> [tool.black]
> line-length = 80
> ```
> 
> And it worked for me.

I think that's pretty new then. We tried this back during the original discussion without luck. But feel free to either update the docs to reflect the reality or post to discourse suggesting a policy change. 

I have no strong feelings either way as long as we don't have to remember to manually change options. A change here would also mean a lot of reformatting which is a bit unnecessary churn.

jdenny-ornl commented:
> But feel free to either update the docs to reflect the reality or post to discourse suggesting a policy change.

I'll plan to raise it in the [original RFC](https://discourse.llvm.org/t/rfc-document-and-standardize-python-code-style/68257) before proposing any changes.

Thanks for all your work in this area.  It is much appreciated.

jdenny-ornl commented:
I'm not trying to rush reviewers, but I want it to be clear that this patch is still in need of review due to recent changes.   See <https://github.com/llvm/llvm-project/pull/65242#issuecomment-1712676799> for a summary.  I'm afraid the recent discussion of black might have buried that comment.  Maybe I'm just not used to LLVM PR reviews yet.

RoboTux commented:
> I'm not trying to rush reviewers, but I want it to be clear that this patch is still in need of review due to recent changes. See [#65242 (comment)](https://github.com/llvm/llvm-project/pull/65242#issuecomment-1712676799) for a summary. I'm afraid the recent discussion of black might have buried that comment. Maybe I'm just not used to LLVM PR reviews yet.

I saw it but was waiting on the updated version fixing the python issue before giving it another look. I shall look at it today.

RoboTux commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -1837,13 +1830,7 @@ def _handleCommand(cls, line_number, line, output, keyword):
-                keyword=keyword, line_number=line_number
-            )
-            assert re.match(
-                kPdbgRegex + "$", pdbg
```
Why is the $ dropped in the buildPdbgCommand?

jdenny-ornl replied:
Good question.

Originally, my goal was for the assert to mimic how kPdbgRegex will actually be used later when expanding `%dbg` substitutions.  That's the usage that the assert is trying to verify will work correctly.  Thus, this patch changes the assert in two ways: the searched string contains all of `%dbg(...) cmd-line` instead of just `%dbg(...)`, and it does not use `$`.

However, now that you ask, I think the assert should be stricter.  For example, [D154987](https://reviews.llvm.org/D154987) proposes to extend kPdbgRegex to permit newlines, which might appear in `lit.run(cmd)` in PYTHON directives, as in the example in that review's summary.  Using `$` in the assert here would have caught that existing deficiency of kPdbgRegex: its `.*` doesn't match beyond the first newline.

Instead of `$`, what do you think of using `re.fullmatch` in the assert?  That is, buildPdbgCommand expects kPdbgRegex to match the entire string it's building.

jdenny-ornl replied:
I went ahead and applied that change and wrote a commit log to explain it.

DavidSpickett replied:
At least from the web UI you have the chance to edit the final commit message. So you can combine them there if you're ok with it all being one commit.

(otherwise yes manually push some of it)

jdenny-ornl replied:
Thanks for the review!

Yes, I performed most steps from the git command line: squash commits and commit logs together, rebase onto the latest main, and force push to the PR.  I then clicked "Squash and merge" and copied and pasted my new commit log there.  The commit log it initially offered was the original comment I posted for this PR, but the focus of the PR has shifted since then.  It would be nicer if it offered the current commit logs, squashed in the same way a `git merge --squash` would squash them.

I don't know how to accomplish the rebase onto main via the github web UI... even though it seems like that's almost always going to be required given how frequently new commits show up on main.  That means I needed to work from the command line anyway, so it seemed more practical to handle squashing there too.

Please let me know if I've overlooked best practices for using the github web UI.

RoboTux replied:
From https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits it seems squash and merge is actually a squash and rebase (it makes it clear it's a fast forward).

pogo59 replied:
The web UI squash-and-merge will automatically rebase to current HEAD.

jdenny-ornl replied:
Thanks for the clarifications. For some reason, I thought it previously rejected one when I wasn't caught up to main, but perhaps I had conflicts then.  I'll try again next time.

jdenny-ornl replied:
Yep, it worked fine another PR that had no conflicts.

RoboTux reviewed (approved):
LGTM

jdenny-ornl committed:
[lit] Fix some issues from --per-test-coverage

D154280 (landed in 64d19542e78a in July, 2023) implements
`--per-test-coverage` (which can also be specified via
`lit_config.per_test_coverage`).  However, it has a few issues, which
the current patch addresses:

1. D154280 implements `--per-test-coverage` only for the case that lit
   is configured to use an external shell.  The current patch extends
   the implementation to lit's internal shell.

2. In the case that lit is configured to use an external shell,
   regardless of whether `--per-test-coverage` is actually specified,
   D154280 causes `%dbg(RUN: at line N)` to be expanded in RUN lines
   early and in a manner that is specific to sh-like shells.  As a
   result, later code in lit that expands it in a shell-specific
   manner is useless as there's nothing left to expand.  The current
   patch cleans up the implementation to avoid useless code.

3. Because of issue 2, D154280 corrupts support for windows `cmd` as
   an external shell (effectively comments out all RUN lines with
   `:`).  The current patch happens to fix that particular corruption
   by addressing issue 2.  However, D122569 (landed in 1041a9642ba0 in
   April, 2022) had already broken support for windows `cmd` as an
   external shell (discards RUN lines when expanding `%dbg(RUN: at
   line N)`).  The current patch does not attempt to fix that bug.
   For further details, see the PR discussion of the current patch.

The current patch addresses the above issues by implementing
`--per-test-coverage` before selecting the shell (internal or
external) and by leaving `%dbg(RUN: at line N)` unexpanded there.
Thus, it is expanded later in a shell-specific manner, as before
D154280.

This patch introduces `buildPdbgCommand` into lit's implementation to
encapsulate the process of building (or rebuilding in the case of the
`--per-test-coverage` implementation) a full `%dbg(RUN: at line N)
cmd` line and asserting that the result matches `kPdbgRegex`.  It also
cleans up that and all other uses of `kPdbgRegex` to operate on the
full line with `re.fullmatch` not `re.match`.  This change better
reflects the intention in every case, but it is expected to be NFC
because `kPdbgRegex` ends in `.*` and thus avoids the difference
between `re.fullmatch` and `re.match`.  The only caveat is that `.*`
does not match newlines, but RUN lines cannot contain newlines
currently, so this caveat currently shouldn't matter in practice.

The original `--per-test-coverage` implementation avoided accumulating
`export LLVM_PROFILE_FILE={profile}` insertions across retries (due to
`ALLOW_RETRIES`) by skipping the insertion if `%dbg(RUN: at line N)`
was not present and thus had already been expanded.  However, the
current patch makes sure the insertions also happen for commands
without `%dbg(RUN: at line N)`, such as preamble commands or some
commands from other lit test formats.  Thus, the current patch
implements a different mechanism to avoid accumulating those
insertions (see code comments).

xgupta commented:
Thank you @jdenny-ornl for fixing the issues.

jdenny-ornl commented:
> Thank you @jdenny-ornl for fixing the issues.

By the way, thanks for the feature!
